File: src/userlinks.pl
# Copyright 2014 Eli Foster

use warnings;
use diagnostics;
use strict;
use MediaWiki::API;
use SatanicBot::Utils;

my $ERRNO = $!;
my $mwapi = MediaWiki::API->new();
$mwapi->{config}->{api_url} = 'http://ftb.gamepedia.com/api.php';

login();
user();
#talk();
logout();

sub login{
    my @login = SatanicBot::Utils->get_secure_contents();
    $mwapi->login({
        lgname => $login[0],
        lgpassword => $login[1]
    }) or die $mwapi->{error}->{code} . ': ' . $mwapi->{error}->{details};
    return 1;
}

sub user{
    my $file = 'info/list_user.txt';
    open my $fh, '<', $file or die "Could not open $file $ERRNO\n";
    print "Starting loop...\n";
    while (my $line = <$fh>){
        chomp $line;
        my $ref = $mwapi->get_page({title => $line});
        my $text = $ref->{'*'};
        $text =~ s/\{\{U\|SatanicSanta/\{\{U\|TheSatanicSanta/g;

        print "Text and article variables have been set.\n";
        eval {
            $mwapi->edit({
                action => 'edit',
                title  => $line,
                text   => $text,
                bot    => 1,
                minor  => 1
            })
        };
        if ($@) {
            print $mwapi->{error}->{code} . ': ' . $mwapi->{error}->{details};
            continue;
        }
        print "Page \'$line\' has been edited.\n";
    }
    close $fh;
    print 'File closed.';
    return 1;
}

sub talk{
#    my @tlinks = ($mw->what_links_here("User talk:SatanicSanta"));
#
#    foreach (@tlinks){
#        my $talk_ref = $mwapi->get_page({title => $_});
#        my $replace_talk = $talk_ref->{'*'};
#
#        $replace_talk =~ s/\[\[User talk:SatanicSanta\]\]/\[\[User talk:TheSatanicSanta\]\]/;
#    }
}

sub logout{
    $mwapi->logout();
    return 1;
}

File: src/ruby/userlinks.rb
require 'mediawiki_api'
require_relative 'wikiutils'
require_relative 'generalutils'

$mw = MediawikiApi::Client.new('http://ftb.gamepedia.com/api.php')
$mw.log_in(General_Utils::File_Utils.get_secure(0).chomp, General_Utils::File_Utils.get_secure(1).chomp)
$other_mw = Wiki_Utils::Client.new('http://ftb.gamepedia.com/api.php')

# This has not yet been tested.
def edit(page_name)
  decoded_json = JSON.decode($other_mw.get_backlinks(page_name))
  backlinks = decoded_json["query"]["backlinks"] do |title|
    title["title"]
  end
  
  backlinks.each do |wlh|
    JSON.parse($other_mw.get_wikitext(wlh))["query"]["pages"].each do |revid, data |
      $revid = revid
      break
    end
    text = JSON.parse($other_mw.get_wikitext(wlh))["query"]["pages"][$revid]["revisions"][0]["*"]
    text = text.gsub(/\{\{U|SatanicSanta/, "{{U|TheSatanicSanta")
    text = text.gsub(/\[\[[Uu]ser:SatanicSanta/, "[[User:TheSatanicSanta")
    text = text.gsub(/\[\[[Uu]ser talk:SatanicSanta/, "[[User talk:TheSatanicSanta")
    $mw.edit(title: wlh, text: text, bot: 1, summary: "Fixing my master's user links")
    puts wlh + " has been edited.\n"
  end
end

edit("User:SatanicSanta")
edit("User talk:SatanicSanta")
exit
